discord-screenaudio: init at 1.9.2

huantianad opened the pull request:
## Description of changes
https://github.com/maltejur/discord-screenaudio
> A custom discord client that supports streaming with audio on Linux. 

Follow up for #199379, closes #186957. Now properly works since #247726 closed #197728.

Not sure if I'm doing something wrong that means I have to pass those `cmakeFlags` to the pipewire library, or is that a bug in the rohrkabel library's pipewire detection? Also, is there a better way to not have rohrkabel install headers into `$out/include`?

@michaelBelsanti, in case you're still interested.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

huantianad committed:
discord-screenaudio: init at 1.9.2

Astavie commented:
the issue is in rohrkabel I believe for pipewire detection. The thing is that discord-screenaudio uses pipewire 3 but qtwebengine uses pipewire 2. Due to the power of Nix this just works :tm: but it does mean both are loaded in pkg-config and it seems rohrkabel uses pipewire 2 by default, so I think manually setting the path is the correct solution here.

huantianad commented:
I wonder if the pipewire search issue could be fixed if we updated rohrkabel from v1.2 to v1.5? I tried but didn't have much luck.

Also, regarding it installing the headers and library, I've made a quick PR upstream that fixes the issue: https://github.com/maltejur/discord-screenaudio/pull/170. I'm pulling in a patch from that instead of removing the folders in `preFixup`.

Astavie commented:
as far as I can see there is not really a precedence for stating dependency services like pulseaudio/pipewire/xdg-portals for packages, though I do see this as a limitation of nix itself. do correct me if there actually is a standard way of doing that though!

huantianad commented:
Yeah I think we'd have to use a nixos module to make sure those options are enabled. However, maybe we can maybe just make a note in the source and longDescription and that would probably be fine. I don't think it's a big deal anyway, the package is probably mergeable without it

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1087

michaelBelsanti commented:
Glad you were able to get it working! It seems that the url for the patch used is no longer valid, so the build fails atm. Did you delete the repo that contained the patch?

huantianad commented:
:facepalm: my bad, I forgot about that! I'll pull the patch in from the main repo.

huantianad commented:
Updated to latest version, which no longer needs the patch. Also, I added an nix-update-script.

marte26 commented:
Hi, I'm currently trying discord-screenaudio on my NixOS install (nixpkgs-unstable) but I get this error:
`[642999:642999:0910/004959.994352:ERROR:base_capturer_pipewire.cc(279)] Failed to load the PipeWire library and symbols.`
The stream starts for a moment and then stops again by itself.
If you think it's an upstream issue I'll write there.

huantianad commented:
@marte26 that's a strange error, I haven't encountered that before. I'd assume at first it's because the app couldn't find pipewire installed on the system, but I feel that'd give a different error rather than a failed to load library error.

marte26 commented:
@huantianad I'm running PipeWIre on my system and everything is working fine, so I don't think that's the issue.
If you need more info from me just ask, I'd really love to get the app working.

huantianad commented:
hmm looking at the error some more it seems like not an error with rohrkabel, but possibly with qtwebengine's webrtc failing to load pipewire?

huantianad commented:
@marte26 After further investigation, I've found the source of the issue. qtwebengine requires pipewire libraries at runtime in order to use pipewire with webrtc. I've created a patch that adds it to LD_LIBRARY_PATH, and that fixes the issue- sort of.
[patch.txt](https://github.com/NixOS/nixpkgs/files/12623153/patch.txt)
This causes a segfault instead when you try to screenshare. I believe this is because qt5's qtwebengine uses pipewire 0.2 which isn't compatible with the pipewire .3 installed on most systems. However, not entirely sure if this is the case

marte26 commented:
So, @huantianad I tried your patch but unfortunately it doesn't work. It segfaults the instant I start a stream, what did work was compiling the project using Qt6 instead of Qt5 (without the patch). Everything worked smoothly with that so I submitted a small PR upstream (maltejur/discord-screenaudio#181) to compile with Qt6 without any modifications.

Thank you so much for the help!

huantianad commented:
Yep yep makes sense, I was actually messing around with using qt6 to both fix the pipewire path problem and also your issue, good to know that it works. I may switch this package to qt6 since it doesn't work on wayland otherwise.

Astavie commented:
Oo if you get it working on qt6 I'd definitely prefer that!

Astavie commented:
Just tested on my own pkgs flake 

https://github.com/Astavie/astapkgs/blob/main/discord-screenaudio.nix#L1

and qt6 does indeed work! when adding pkg-config to the native build inputs we don't even have to manually set the pipewire path as qt6's qtwebengine uses pipewire 3 as well

huantianad commented:
Yep yep that's another reason I was considering it, the only downside is no KF6 libs in nixpkgs yet

id3v1669 commented:
> Just tested on my own pkgs flake
> 
> https://github.com/Astavie/astapkgs/blob/main/discord-screenaudio.nix#L1
> 
> and qt6 does indeed work! when adding pkg-config to the native build inputs we don't even have to manually set the pipewire path as qt6's qtwebengine uses pipewire 3 as well

Hi, tried using your configuration. Do you experience any issues with it?
I'm asking because I have one: after I close app, I loose session and have to relogin each time.

fin444 commented:
@id3v1669 see https://github.com/ArmCord/ArmCord/issues/407#issuecomment-1710643529

You can see the dev tools by running `QTWEBENGINE_REMOTE_DEBUGGING=someport discord-screenaudio` and then connecting to `localhost:someport` in a browser. Check for the `token` and `tokens` local storage items. If they're missing, then I would guess Discord is recognizing it as a non-browser client on Qt6 but not Qt5. On Flatpak, the tokens are not hidden.

id3v1669 commented:
@fin444 
token and tokens appear in local storage after successful login but on relaunch of app they are just not there.
tried adding NoDevtoolsWarning plugin, still same issue.
Looks like those lines might be related to this issue.
[discord] [ProcessUtilsElectron] setMemoryInformation not available.
[discord] [DatabaseManager] active user changed (now: undefined, was: null, was: undefined)

fin444 commented:
@id3v1669 I also have the `ProcessUtilsElectron` line on Flatpak, so I don't think it's related. For the `DatabaseManager` line, on Flatpak the `now` is set to my Discord numeric ID rather than `undefined` so that could be related. Doing a quick look in the code, that message is generated by `handleAuthenticationStoreChanged()`, so seems right on the money. But, trying to understand this obfuscated code is a bit beyond my abilities so I'm not really sure what this means.

huantianad commented:
Just updated this package to 1.9.1 and also switched it to qt6 to use pipewire 0.3 for wayland.

I'm also experiencing the app logging me out each time when I close it. Funnily enough, enabling the vencord plugin also doesn't work, but it seems that the plugin is also turned off on restart, and it seemingly doesn't save. Perhaps this is something to report to upstream?

heyimnova commented:
Hi, I just tried out this package on NixOS. Streaming is broken on 1.9.1 however updating to 1.9.2 fixes streaming with this package.
Streaming a second monitor on Wayland is buggy (low framerate/not refreshing parts of the stream) however adding qtwayland to buildInputs fixes that issue :)

huantianad commented:
Thanks for the feedback! I'll add those changes soon!

Artturin commented on pkgs/applications/networking/instant-messengers/discord-screenaudio/default.nix:
```diff
@@ -0,0 +1,56 @@
+    cmake
+    pkg-config
+    qtbase
+    qtwebengine
```
This should be in buildInputs if no binaries from it are used

huantianad replied:
Referring to the two qt packages here? Is that the rule in general? I always thought that it was if a package isn't needed at runtime, it should be put in nativeBuildInputs.

Artturin replied:
Build time tools are in nativeBuildInputs

Libs and packages to be added to the wrapper (if in use) should be in buildInputs 

Add a `strictDeps = true` to separate the dependencies better and not add buildInputs to the build PATH

Artturin replied:
It's good to use it always, note that some ecosystem specific builders such as `buildRustPackage` and `buildPythonApplication` always enable it.

Check with 
```
$ nix eval ".#bash.strictDeps"
true
```
when in doubt.
